chapel-lang/chapel
# Move C Types into a new CTypes module

As part of stabilizing the standard modules, this PR moves our C type
declarations from SysBasic, SysCTypes, and CPtr into a single, new
CTypes module, deprecating the previous modules when empty, while
keeping them working for the time being.  In the case of SysBasic, I
moved some, but not all, of its contents into other modules as a step
toward retiring it, leaving primarily things related to system errors
(which seem like likely candidates for SysError in future work, though
I'm not as familiar with that module).  I view the next step for SysBasic
to be for it to be reviewed as part of the module stabilization process,
ideally simultaneously with, or adjacent to, SysError.

Note that this PR does not implement everything that we hope to do to
stabilize the `CTypes` module, it merely moves the major pieces and
components into a separate module.  Follow-up issues will deal with
other issues like:

- [ ] https://github.com/chapel-lang/chapel/issues/18014
- [ ] https://github.com/chapel-lang/chapel/issues/18015
- [ ] https://github.com/chapel-lang/chapel/issues/18011
- [ ] https://github.com/chapel-lang/chapel/issues/18012
- [ ] Are we happy with the name `isAnyCPtr()`?
- [ ] And other things from the module review...

In more detail, it:
* creates a new CTypes module that contains:
  * all of ChapelSysCTypes via a `public use`
  * all of CPtr's contents, which were copied over
  * `c_float` and `c_double`, copied over from SysBasic
* changes the name of the generated SysCTypes module to
  ChapelSysCTypes in order to support a user-facing SysCTypes
  module that can contain the deprecation warning and
  documentation while forwarding to ChapelSysCTypes
* deprecates CPtr, replacing its contents with a `public use` of the
   parts of `CTypes` that reflect what it used to contain.  Note that
   this is currently preserving behavior only because of the compiler's
   mishandling of operator methods and public use statements.  See
   issue #19256 for more on this.
* updates SysBasic as follows:
  * moves `c_float`, `c_double` to CTypes, replacing them with procs
    that document the move
  * moves socklen_t, mode_t, off_t to Sys.chpl since that's where
    they're used, replacing them with procs that document the move
  * moves chpl_cnullfile(), chpl_cstdout(), chpl_cstderr() from
    SysBasic to IO since those are the only place they're used and
    they're not user-facing 
* updates the chpldocumentation for all four modules to
  indicate the deprecations and revised contents

Looking at the roles of various files:

* update chpldoc logic to special-case ChapelSysCTypes and
  make it renderable:
  * compiler/AST/ModuleSymbol.cpp
  * modules/Makefile
  * util/config/make_sys_basic_types.py
* add new tests making sure deprecation / move messages
  work as intended:
  * test/deprecated/*
* move contents of SysBasic to IO, Sys, and CTypes:
  * modules/standard/SysBasic.chpl
  * modules/standard/IO.chpl
  * modules/standard/Sys.chpl
* update the compiler to use CTypes for extern features rather
   than the other three modules:
   * compiler/AST/build.cpp
   * compiler/passes/expandExternArrayCalls.cpp
* update c2chapel to use CTypes:
  * tools/c2chapel/c2chapel.py
* update the compiler to check for ChapelSysCTypes rather
   than SysCTypes as part of its sanity checks:
   * compiler/parser/parser.cpp
* update documentation to avoid referring to old modules:
  * doc/rst/technotes/extern.rst
* add CTypes to the list of standard modules to render:
  * doc/rst/meta/modules/standard.rst
* update modules to only refer to CTypes rather than old ones:
  * modules/*
* update tests as follows:
  * test/compflags/ferguson/print-module-resolution.good: reorder
    modules to reflect new resolution order 
  * test/modules/bradc/printModStuff/foo.good: similar
  * test/modules/sungeun/init/printModuleInitOrder.good
  * test/modules/sungeun/init/printModuleInitOrder.na-none.good
  * test/release/examples/primers/interopWithC.chpl: to reflect new
    modules
  * test/types/records/ferguson/self-pointer-record/myrecord.chpl:
    updated (skipped) test to reflect current Chapel
  * test/* : everything else is updated to use CTypes rather than old
     modules

Resolves #18013
